### DISCUSSION TITLE
Deployment Models - Design discussion

### DISCUSSION THREAD

**Discussion by lucasponce** [kiali/kiali#4081] [Roadmap]
Hi everyone,

The aim of this discussion is to consolidate all the brainstorming that may help into the design of the support of multiple Istio deployment models in Kiali [1] [2].

This topic is very active, and it has already had preliminary discussions in the community [3] https://github.com/kiali/kiali/discussions/3952 .

Kiali has support for multi primary scenarios [4] but more models need to be supported (primary/remote) and that opens more architecture challenges that may impact in the design.

Please, join us in these discussions :)

[1] https://istio.io/latest/docs/ops/deployment/deployment-models/
[2] https://istio.io/latest/docs/ops/deployment/deployment-models/#control-plane-models
[3] https://github.com/kiali/kiali/discussions/categories/multicluster
[4] https://kiali.io/documentation/latest/features/#_advanced_mesh_deployment_and_multi_cluster_support

**Comment by lucasponce** [comment]
One of the ideas that have been discussed within the team is the need to introduce a "explicit", "control plane" view in Kiali.

Today, Kiali assumes that there is a single control plane, and there are some "implicit" management in the "control plane" status health (mast head), proxy health and other internal features as using the Istio registry.

In Multi-Cluster scenarios Kiali has introduced an experimental feature where a "Mesh" page defines the "configuration" of the control planes.

But I think this idea needs to evolve into a page that list the "logical" control planes involved, that would help in Deployment Models and this "logical entity" could be something similar as the "Application" is for the data plane.

In a "Mesh/Control Plane" page I would expect to see:
- List of control planes that Kiali has visibility (it could be in a single cluster or different clusters).
- Status of the control plane (pretty similar to the health status that today Kiali offers in the mast head but per listed elements).
- Access to the config relevant of the control plane (close to what the Mesh page shows today).
- Access to the status of the control plane (a view to dump registry and security views).

In the "Data Plane" (Workloads) Kiali should show:
- If a workload is connected to control plane A or B (remember this is orthogonal to deployment models, in a canary model this could be a normal situation).
- As it happens today, "Data Plane" should enrich the "Workload" status with runtime information from the control plane (Today it shows the sync with the control plane, but there are more feature to address: status of security, probably this could be added into the new "Envoy" tab that is being worked).

**Reply by spockz** [reply]
I agree with the idea as written in the comment above but do have some implementation questions.
How would we implement retrieving the data from the remote clusters? If we integrate on the kubeApi or istiod level we would need rpc level interaction between the "governance" cluster running kiali and all the remote clusters. 

In our setup we have multiple clusters, each running their own IstioD control plane. The clusters and istiod form one single logical mesh where the remote clusters/workloads are configured into each cluster by our governance process using events from the individual clusters. 

Would we have the ability to point Kiali to a custom service which hosts our aggregated view, scoped per cluster/istio install? Or would Kiali expect to be calling the kubeapi of the remote cluster, or a "kiali-agent"?

**Reply by lucasponce** [reply]
Thanks @spockz, indeed is a good question that needs to be picture on this discussion.

I guess that a first would be to mimic similar pattern used by istiod on [1], so, if "istiod" connects to a remote cluster and it has access to the kubernetes api, Kiali should follow that to not introduce more elements to manage.

But this is in an early discussion, so all options are open.

[1] https://istio.io/latest/docs/ops/deployment/deployment-models/#control-plane-models

**Reply by osswangxining** [reply]
Alibaba Cloud Service Mesh (ASM, https://servicemesh.console.aliyun.com) provides one hosted control plane running on managed cloud side to manage one or more K8s clusters as data plane.  From the feedback of most customers, we think that it's first priority to make Kiali to support this separated deploy model - split control plane and data plane

**Reply by stu233win** [reply]
A cloud vendor’s managed control plane is a typical example of an external control plane.
[1] https://istio.io/latest/docs/ops/deployment/deployment-models/#control-plane-models
[2] https://www.alibabacloud.com/help/doc-detail/200336.htm

**Reply by jshaughn** [reply]
@osswangxining So, as a provider, is it true that the different data planes do not talk with each other?  That each cluster is a "tenant" for a particular customer?

**Reply by lucasponce** [reply]
Also one interesting question @stu233win @osswangxining could be to learn how telemetry is typically collected on this external control plane strategy (i.e. one prometheus server per cluster or one prometheus server per control plane talking with mulitple clusters, or other design).

**Reply by osswangxining** [reply]
> @osswangxining So, as a provider, is it true that the different data planes do not talk with each other? That each cluster is a "tenant" for a particular customer?

Most of our customer cases use "flatten" network among the different data planes, e.g. multiple K8s clusters, or virtual machines. These infrastructures are using same VPC/private network, or using some network products (such as Cloud Enterprise Network in Alibaba Cloud) to keep them inter-connection.  So these different data planes can talk with each other instead of NOT. All of them belong to one customer.

For different customer/tenant, they will get one different ASM instance - one hosted control plane.

**Reply by osswangxining** [reply]
> Also one interesting question @stu233win @osswangxining could be to learn how telemetry is typically collected on this external control plane strategy (i.e. one prometheus server per cluster or one prometheus server per control plane talking with mulitple clusters, or other design).


From our customers cases, I have the following summary FYI:
- #1  only 1 K8s cluster within data plane (>80% cases, so we think this one should be 1st priority), so naturally one prometheus server for it - but we need to cover 2 options: one is that the prometheus server is running within this K8s cluster, the other is that the prometheus server is running out of this K8s cluster, e.g. one hosted prometheus server, etc. 
- #2 for multiple K8s clusters within data plane, we hope that one prometheus server can collect all of them, but  it depends on the prometheus capabilities. So far no offering for now to easily build the prometheus for multiple clusters.

**Reply by osswangxining** [reply]
@lucasponce I think that we can start from the deploy model - one external/hosted control plane manages one K8s cluster in the data plane.  Kiali dashboard can provide the customization, e.g. only show the Kiai UI parts for control plane, or only show for data plane related artifacts.

**Reply by lucasponce** [reply]
> only 1 K8s cluster within data plane (>80% cases, so we think this one should be 1st priority), so naturally one prometheus server for it - but we need to cover 2 options: one is that the prometheus server is running within this K8s cluster, the other is that the prometheus server is running out of this K8s cluster, e.g. one hosted prometheus server, etc.

One question @osswangxining for this (1) scenario, you'd like to integrate all clusters in a single Kiali, or having a specific Kiali per cluster ? (or to have a flexible option covering both)

**Reply by osswangxining** [reply]
@lucasponce  prefer to one single Kiali to get the overall view. I know this depends on one Prometheus  server or several Prometheus servers for Kiali, it should not easy.

**Comment by jshaughn** [comment]
Today the Kiali deployment model supports the Istio multi-primary deployment model.   Kiali is then deployed on the same primary cluster, configured for the local istiod and for a Prometheus instance scraping metrics only for the local pods.  Kiali can then assume that it has access to all of the services/workloads/config/etc, found in the metrics, as long as the cluster label matches the local cluster.  For source or destinations labeled with a different cluster, it is assumed the user needs to view that config via another Kiali instance, local to the cluster in question.

For a more flexible deployment model we need to be able to remove that local-cluster assumption and more flexibly be able to determine which config is owned/managed by (visible to) the Kiali instance.  This goes back to the mesh/control-plane view above, that identifies all of the control planes configured for the Kiali instance.   (Note, it may also be possible for that view to discover other Kiali's, similar to today, and to offer links to those Kiali instances).  With this sort of configuration Kiali could then better handle unified metrics, a Prometheus instance that combined metrics from different clusters, whether by scraping to a single instance, or using Prometheus federation to combine metrics from multiple Prom instances into a single instance.

So, if a single Kiali were configured for multiple Istio control planes, and could also access any remote clusters for those control planes, it could create a model to support that unified view, and handle unified metrics from a single Prometheus.   Less desirable would be unified control planes without a unified Prometheus.

I suppose it could be feasible to have a Kiali configured for multiple Prom dbs, with only a single Prom db active at any time.  That would give unified config but but not unified metrics.  And overall may seem a little clunky.  I think it is out of scope for Kiali to itself try and unify multiple Prom instances.  That is what Prom federation is there to do, or to perhaps use something like Thanos for unification.

**Reply by lucasponce** [reply]
I don't agree with this.
It's included in the scope of Kiali to support multiple Prometheus instances.

It's out of the scope to force a user to setup a predefined architecture using federation or Thanos.

If a customer is defining a topology where multiple prometheus are involved, Kiali can query them in the same way that it will query mutliple clusters.

**Reply by lucasponce** [reply]
Indeed, Kiali has always supported to fetch metrics from multiple prometheus instances where customers could separate istio/infrastructure metrics from application metrics.

So, this is not a different use case but an implementation detail.

**Reply by jshaughn** [reply]
Well, I'd like to hear from some others on this one.  Maybe I don't quite understand the need to stitch together istio metrics from multiple control planes at the Kiali level.  Open to more input and opinions here.

**Reply by spockz** [reply]
>It's included in the scope of Kiali to support multiple Prometheus instances.
>It's out of the scope to force a user to setup a predefined architecture using federation or Thanos.
> If a customer is defining a topology where multiple prometheus are involved, Kiali can query them in the same way that it will query mutliple clusters.

From a metrics point of view, what is the difference between Kiali scraping multiple prometheus instances (in different clusters and representing different (but optionally "joined") control planes), and having something like Prometheus federation and/or Thanos in place? AFAICT, it would still lack the meta-data that it can deduce from the Istio configuration (the virtual services, routes, etc.)

I would argue that for the metrics part it is even easier to let Prometheus or Thanos do the federation. Just as Kiali is not providing its own tracing implementation.

**Reply by lucasponce** [reply]
The Kiali model will be populated from multiple control planes, not only metrics.
Workload data, Istiod registry, proxy-status and metrics.
I (yes it's my personal opinion/view) think that the model should be flexible enough to fetch data from multiple sources.

Today it's something that Kiali performs with application metrics.
If user has a single prometheus for all control planes, good, if they have some, because they are starting to evaluate, or that's their decision, Kiali model should support multiple prometheus sources (even, if it's not ideal) but I wouldn't like to tie a design to a such architecture decision.

In short, a model with multiple sources can support having one single source a case for multiple control planes/clusters, but I wouldn't constraint to have a single (federated or thanos) instance for all architecture. Some users will have it, others not yet.

Also, I'd keep thinking in the Kiali model, where metrics is just a source to enrich the control plane topology, with the progress of unified standars as OpenTelemetry, I like to think in an abstraction that can be populated, today metrics are used to stablish some relations that tomorrow can be provided from other source.

**Reply by lucasponce** [reply]
> From a metrics point of view, what is the difference between Kiali scraping multiple prometheus instances (in different clusters and representing different (but optionally "joined") control planes), and having something like Prometheus federation and/or Thanos in place? AFAICT, it would still lack the meta-data that it can deduce from the Istio configuration (the virtual services, routes, etc.)
> 
> I would argue that for the metrics part it is even easier to let Prometheus or Thanos do the federation. Just as Kiali is not providing its own tracing implementation.

Both should be supported, the key is to have a model that doesn't force to have Federated Prometheus or Thanos as a mandatory requeriment for a multi cluster architecture.

**Reply by spockz** [reply]
Okay, that confirms my intuition that these are equivalent. In terms of data flow we have "solved" the metrics part. Either users can federate from the different Prometheus instances from Kiali or federate with their own mechanism.

 I'm especially interested in seeing how the "Workload data, Istiod registry, proxy-status" is going to be gathered. Can we implement a similar aggregation/federation as we have for metrics?

**Reply by lucasponce** [reply]
> I'm especially interested in seeing how the "Workload data, Istiod registry, proxy-status" is going to be gathered. Can we implement a similar aggregation/federation as we have for metrics?

I think similar as followed in the deployment models [1] reference.
Kiali should get access to the control plane (one per cluster, one per whole mesh, or multiple combinations) and to the cluster apis.

What it seems doable is that at scale there should be a tradeoff managed by the user between real-time vs scalability.
The more info you get, probably cache techniques should increase the validity of the info in a larger window.

[1] https://istio.io/latest/docs/ops/deployment/deployment-models/

**Reply by jmazzitelli** [reply]
Two things:

If we are going to aggregate metrics from multiple Prometheus instances into this "Kiali model" - it means we will be doing essentially what Prometheus federation/thanos is doing - it means we most likely are going to be required to implement our own storage model and perhaps even a query language (or we use a DB storage/query language). The alternative would be to stitch the results of N queries to N prometheus servers - but that will likely be very expensive as N gets bigger.

Second, if we are to have this Kiali model that somehow aggregates multiple sources of metrics (and istio config), we need to start thinking hard about "where" this model's data is going to be -- Kiali is most likely going to need backend storage. And this is going to be a huge paradigm shift for Kiali - so we are going to need to start thinking about the design for this soon because it will be a big change. What kind of backend storage? How will it be managed? How do we query the data to support all the things we need?

I realize this is "implementation details" - but to me as soon as we start talking about a "Kiali model" that aggregates across multiple clusters/prometheuses, backend storage is going to be required to be baked into the design (unless, again, we think we can efficiently query and stitch together results from N prometheus servers (or N istiod/control planes) within each request for kiali model data).

**Reply by lucasponce** [reply]
If a Istio user have 3 clusters and he/she decides to install a prometheus instance on each cluster, he/she will be able to have a single Grafana to visualize it but not Kiali if we force to have an unified federated solution.

I wouldn't like to add that constraint, Kiali can fetch Kiali information from those 3 clusters (it's going to do it for the control plane and the cluster, doing for the prometheus instance is more an implementation detail).

> Second, if we are to have this Kiali model that somehow aggregates multiple sources of metrics (and istio config), we need to start thinking hard about "where" this model's data is going to be -- Kiali is most likely going to need backend storage.

I think this would be driven by the model; it should start from memory, as it works today, with the potential to be upgraded if required. But +1, it's something it should be on the design board.

As commented, the discussion is about to support both options: if user has a federated solution, perfect; if user has not yet defined one and they want to use individual sources, great. In any case, it will require also to iterate from multiple sources of control planes and clusters.

**Reply by israel-hdez** [reply]
In general, I do not agree that this is the same use case that we currently have. Currently, Kiali only supports two Prometheuses: a main one. Optionally, a second one is supported for the [Custom Dashboards feature](https://kiali.io/documentation/latest/runtimes-monitoring/). So, I see the idea is to let the user decide between using one p8s for everything, or use two (one for istio metrics and one for app metrics).

Adding support for any number of Prometheuses for Istio metrics sounds like a different beast, IMO.

If we want to generalize into a single use-case, we probably should make no distinction between prometheus for Istio metrics and prometheus for custom dashboards, and Kiali should federate metrics from any p8s without any distinction (i.e, assume that any p8s can both have metrics for Istio and metrics for apps and unify all of them for later querying).

Honestly, although I partially like the idea of implementing the federation within Kiali, I reckon the problem is already solved if we use the already available federation feature of Prometheus. Thus, our own implementation of metrics federation does not sound that valuable to become a priority. If there is already a solution, under the context of this "Deployment models" discussion it becomes a secondary problem (Supporting multiple p8s instances adds flexibility to the end user, but that's it.).

I see this subject too distracting and is not allowing us to make progress on the "Deployments model" issue (aka clustering/federation). I do not see how this is going to help us making progress on the "Deployments model" discussion. So, I propose to move the multi-p8s support to a different discussion and ask feedback from community in that separate discussion. I think the needed changes in Kiali to support multi-p8s are big enough to deserve it's own discussion. That other thread can be closely related to this one, similarly to the [Scalability discussion](https://github.com/kiali/kiali/discussions/4080).

**Reply by lucasponce** [reply]
@israel-hdez not sure if I share same vision that this subject is not related to the "Deployment Models" discussion.

Kiali design should work for any deployment model, and to be flexible it would require that can accept configurations of 4 elements:

- Clusters (Data planes): url for the kubernetes API, credentials, cluster identification, etc.
- Control Planes: Which control plane (external or internal) is linked with which cluster.
- Prometheus: Which prometheus instance is linked with which cluster (as discussed, it can be one, or it can be a combination, i.e. I have a federated for 10 cluster, but I add a new one to the mesh with it's own instance), but I need a list identifying which instance is linked with which cluster.
- Tracing: Same as Prometheus, ideally a single instance would be good but, it may have multiple.

**Reply by nrfox** [reply]
Can this be done in phases? Something like

Phase 1: Kiali can run in separate controlplane cluster fetching istio config from kube api in data plane cluster and querying prometheus in data plane cluster.
Phase 2: Kiali can fetch istio config from N data planes but relies on a single federated prometheus
Phase 3: Kiali can fetch istio config from N data planes and N prometheus

Starting with phase 1 today does not prevent you from doing phase 3 in the future. In fact you solve for problems needed later on as you go.

**Comment by israel-hdez** [comment]
Regarding namespace selection, Kiali has the following menu in the graph:

![image](https://user-images.githubusercontent.com/23639005/123881004-69545680-d909-11eb-9608-08ea6ece476a.png)

A similar selector is available in other Kiali pages.

Several clusters can be part of the same Istio Mesh. Across clusters, Istio treats namespaces with same name as the same logical namespace. So, if there is an Istio CRD targeting certain namespace, Istio can apply it's configuration across all interconnected clusters.

However, in some cases there is some data sharding. For example, metrics in Prometheus have labels to identify source and destination clusters of traffic. 

So, I wonder what would be more useful for users in clustered/federated envs?:
* keeping the namespace selector simple like it is right now?
* or be specific and list namespaces per cluster?

The advantage of the former, is that Kiali would offer the logical view that Istio works with. The advantage of the latter is that data can be filtered (for example, you may want to see a graph for only certain cluster? and similarly when listing apps/workloads/services?).

**Reply by lucasponce** [reply]
I have the impression that a selector for clusters would be ideal: it can simulate a "select all clusters" creating that "virtual" namespace across clusters, but also selecting/filtering by clusters would help. So with that component I guess it can define both scenarios that you propose.

**Reply by abonas** [reply]
What about the overview page? it could potentially also display what clusters are part of each namespace (in the visual "cards")

**Reply by israel-hdez** [reply]
@abonas For the overview page, I was thinking it on the other way: show which namespaces belong to what clusters. 

Although if we follow the Istio philosophy about that the mesh is on top of clusters, your suggestion makes more sense.

**Comment by jshaughn** [comment]
The new federated-travel demo is nice and shows mesh federation on a single k8s cluster.  It does leave me with a question:  What exactly do we mean by "cluster" when we talk about multi-cluster, multi-controlPlane or federation?   In the federated travel demo we have two meshes but do we expect one, or two "clusters" in the resulting visualization?   If you enable Cluster Boxes you'll see that both meshes report telemetry as the `Kubernetes` cluster.  `Kubernetes` is the default cluster name assigned by Istio if you don't set values.global.multiCluster.clusterName at install-time.

![image (1)](https://user-images.githubusercontent.com/2104052/142487960-b03bc7e6-2c7a-4cfc-a17a-1a22a8109faf.png)

As-is, the only way to know requests are traveling across federated meshes is by the mesh-based namespace naming used in the demo.  And that may be fine, because there truly is no traffic crossing k8s clusters.  As an aside, I feel like, though, that OSSM should probably assign a non-default cluster name to each federated mesh, even if it were the same in this single-cluster scenario.

Other than the possible visualization issue, none of this is really a problem; each control plane has a Kiali instance and because the control planes don't share a namespace, it is clear which workloads belong to each control plane and Kiali knows what is local, accessible config and what is "restricted" (i.e. inaccessible/not found).

But if we are looking towards a single Kiali that can work against multiple control planes (initially on the same cluster), and the unified metrics of each control plane, we need to be able to work backwards from metrics to registryEntries/config.  If namespaces are shared we need some other sort of discriminator and some sort of Kiali configuration to help us register and describe the scope of each control plane.

**Reply by lucasponce** [reply]
I think this is related to the make the "Control planes" explicit idea [1].

As it is pointed there may be multiple combinations here:
- Single cluster: multiple control planes in the same mesh (canary upgrades).
- Single cluster: multiple control planes in different meshes (federated/non-federated).
- Multi cluster: a single control plane/mesh deployed in different clusters.
- Multi cluster: multiple meshes separated in cluster but talking on each one.

In upstream perhaps all scenarios may happen, and in products like OSSM there could be more restricted (canary upgrades are not supported in Maistra).

So, I tend to think that probably we need to build the information looking first into the registry (what are the elements in my mesh ?) and potentially complete the model with telemetry (connections, status), and that will define the boundaries of the logical units that would be interesting to highlight.

[1] https://github.com/kiali/kiali/issues/4435

**Reply by nrfox** [reply]
I don't think Kiali should guess at what a cluster is and should just take what gets reported by metrics as the source of truth. If two clusters have the same name but are actually different, that is a configuration issue and users can fix that by properly setting `global.multiCluster.clusterName` to different values. For what it's worth, there's some [early stage Kube proposals](https://github.com/kubernetes/enhancements/tree/master/keps/sig-multicluster/2149-clusterid) to assign unique ids to clusters so perhaps in the future, the default `global.multiCluster.clusterName` would be assigned to this instead of `Kubernetes` like it is today.

**Reply by lucasponce** [reply]
I'm +1 to avoid "auto-discovery" tasks here and let the user to define these boundaries through configuration (define what a cluster and a mesh is).

**Reply by jshaughn** [reply]
No argument from me, I don't want Kiali involved with the configuration of the mesh(es),  but I am wondering about what Kiali will need in its own config to do its job.  My question is about how we "map" an entry in telemetry to its "owning" control plane when we are dealing with unified metrics and Kiali is managing multiple control planes.  Note also, I think, that the telemetry entry may be outside of those managed control planes.
